From 483bceb24f3e8a29b159d1cf461a4b0894cc0d20 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Mon, 3 Mar 2025 15:37:19 +0800
Subject: [PATCH 181/210] mtk: hostapd: rework add txpower vendor command

Set default sku index as -1 and hostapd won't bring sku index, if
the config didn't specify the index larger than 0.

The meaning of sku index is the same as usually, 0 is don't specify
the index, larger than 0 indicates driver need to match the
corresponding index in dts.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 src/ap/ap_config.c           | 4 ++--
 src/ap/ap_config.h           | 4 ++--
 src/ap/ap_drv_ops.c          | 2 +-
 src/drivers/driver.h         | 4 ++--
 src/drivers/driver_nl80211.c | 7 ++++---
 5 files changed, 11 insertions(+), 10 deletions(-)

diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index f636b43d3..4838098f9 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -318,8 +318,8 @@ struct hostapd_config * hostapd_config_defaults(void)
 	conf->pp_mode = PP_FW_MODE;
 
 	conf->lpi_psd = 0;
-	conf->sku_idx = 0;
-	conf->lpi_sku_idx = 0;
+	conf->sku_idx = -1;
+	conf->lpi_sku_idx = -1;
 	conf->lpi_bcn_enhance = 0;
 
 	hostapd_set_and_check_bw320_offset(conf, 0);
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 0ba32ad52..603c42f50 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1370,8 +1370,8 @@ struct hostapd_config {
 	void *muru_config;
 	u8 pp_mode;
 	u8 lpi_psd;
-	u8 sku_idx;
-	u8 lpi_sku_idx;
+	s8 sku_idx;
+	s8 lpi_sku_idx;
 	u8 lpi_bcn_enhance;
 };
 
diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 80a2dbeb8..ed38c6ad1 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1558,7 +1558,7 @@ int hostapd_drv_txpower_ctrl(struct hostapd_data *hapd)
 			goto out;
 	}
 
-	if (hapd->iface->afc.lpi_mode == true)
+	if (hapd->iface->afc.lpi_mode == true && hapd->iconf->lpi_sku_idx > -1)
 		sku_idx = hapd->iconf->lpi_sku_idx;
 #endif /* CONFIG_AFC */
 
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index f0bcaad97..bb9760079 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5712,8 +5712,8 @@ struct wpa_driver_ops {
 	* @power_table: power table generated from AFC response
 	* @lpi_mode: specify the current mode is whether lpi
 	*/
-	int (*txpower_ctrl)(void *priv, u8 lpi_psd, u8 sku_idx, u8 lpi_bcn_enhance,
-			    u8 link_id, s8 **power_table, u8 lpi_mode);
+	int (*txpower_ctrl)(void *priv, u8 lpi_psd, s8 sku_idx, u8 lpi_bcn_enhance,
+			    s8 link_id, s8 **power_table, u8 lpi_mode);
 };
 
 /**
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index f285b43f8..4939f9069 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -16587,8 +16587,8 @@ fail:
 	return -ENOBUFS;
 }
 
-static int nl80211_txpower_ctrl(void *priv, u8 lpi_psd, u8 sku_idx, u8 lpi_bcn_enhance,
-				u8 link_id, s8 **power_table, u8 lpi_mode)
+static int nl80211_txpower_ctrl(void *priv, u8 lpi_psd, s8 sku_idx, u8 lpi_bcn_enhance,
+				s8 link_id, s8 **power_table, u8 lpi_mode)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
@@ -16617,7 +16617,8 @@ static int nl80211_txpower_ctrl(void *priv, u8 lpi_psd, u8 sku_idx, u8 lpi_bcn_e
 		goto fail;
 
 	nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD, lpi_psd);
-	nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX, sku_idx);
+	if (sku_idx > -1)
+		nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX, sku_idx);
 	nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE, lpi_bcn_enhance);
 
 	if (link_id > -1)
-- 
2.45.2

